src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,169 @@
+            ? '/' + componentBehindPath
+            : componentBehindPath;
+
+        var componentBehindUri = new UriBuilder
+        {
+            Scheme = Uri.UriSchemeFile,
+            Path = updatedComponentBehindPath,
+            Host = string.Empty,
+        }.Uri;
+
+        var text = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (text is null)
+        {
+            return null;
+        }
+
+        var componentName = Path.GetFileNameWithoutExtension(componentBehindPath);
+        var componentBehindContent = text.GetSubTextString(new CodeAnalysis.Text.TextSpan(actionParams.ExtractStart, actionParams.ExtractEnd - actionParams.ExtractStart)).Trim();
+
+        var start = componentDocument.Source.Text.Lines.GetLinePosition(actionParams.ExtractStart);
+        var end = componentDocument.Source.Text.Lines.GetLinePosition(actionParams.ExtractEnd);
+        var removeRange = new Range
+        {
+            Start = new Position(start.Line, start.Character),
+            End = new Position(end.Line, end.Character)
+        };
+
+        var componentDocumentIdentifier = new OptionalVersionedTextDocumentIdentifier { Uri = actionParams.Uri };
+        var componentBehindDocumentIdentifier = new OptionalVersionedTextDocumentIdentifier { Uri = componentBehindUri };
+
+        var documentChanges = new SumType<TextDocumentEdit, CreateFile, RenameFile, DeleteFile>[]
+        {
+            new CreateFile { Uri = componentBehindUri },
+            new TextDocumentEdit
+            {
+                TextDocument = componentDocumentIdentifier,
+                Edits = new[]
+                {
+                    new TextEdit
+                    {
+                        NewText = $"<{componentName} />",
+                        Range = removeRange,
+                    }
+                },
+            },
+            new TextDocumentEdit
+            {
+                TextDocument = componentBehindDocumentIdentifier,
+                Edits  = new[]
+                {
+                    new TextEdit
+                    {
+                        NewText = componentBehindContent,
+                        Range = new Range { Start = new Position(0, 0), End = new Position(0, 0) },
+                    }
+                },
+            }
+        };
+
+        return new WorkspaceEdit
+        {
+            DocumentChanges = documentChanges,
+        };
+    }
+
+    /// <summary>
+    /// Generate a file path with adjacent to our input path that has the
+    /// correct code-behind extension, using numbers to differentiate from
+    /// any collisions.
+    /// </summary>
+    /// <param name="path">The origin file path.</param>
+    /// <returns>A non-existent file path with the same base name and a code-behind extension.</returns>
+    private static string GenerateComponentBehindPath(string path)
```

Comment:
Rather than copy this method from `ExtractToCodeBehindCodeActionResolver`, it would be good to move it to a shared file somewhere, and have both that class and this class use the same method.